### PR TITLE
feat(dyn-sampling): Add "latest" release sampling [TET-355 TET-311]

### PR DESCRIPTION
**Pull request by andriisoldatenko**
This PR adds ability to store "latest" literal in dynamic sampling
rule json, and update it to concrete release version using `sentry.search.utils.get_latest_release`
in project config endpoint, so relay can use only concrete version of release.

TODO:
- Added Trigger recomputation of project_config cache as soon as new `Release` has been created.
- Add unit test for checking `transaction.on_commit`

Also i did few tests:

E.g.
0. Create dynamic sampling rule with "latest" literal.

1. Create release via sending transaction:
```
In [3]: import sentry_sdk
   ...:
   ...: sentry_sdk.init(
   ...:     "http://a4b7b57142ce4fdb8308a7484c2a4250@localhost:8000/1",
   ...:     # Set traces_sample_rate to 1.0 to capture 100%
   ...:     # of transactions for performance monitoring.
   ...:     # We recommend adjusting this value in production.
   ...:     traces_sample_rate=1.0,
   ...:     # environment="staging",
   ...:     release="myapp@1.0.1",
   ...: )
   ...:
   ...:
   ...: with sentry_sdk.start_transaction(name="andrii") as t:
   ...:     t.set_tag("andrii", "test transaction")
   ...:     raise Exception
```

2.  Check UI:
image

3. Send POST to projectconfig endpoint:
```
curl \ 
  -d '{"publicKeys":["***"],"fullConfig":true,"noCache":true}' 
  -H "Content-Type: application/json" 
  -H "X_SENTRY_RELAY_SIGNATURE: ***" 
  -H "X_SENTRY_RELAY_ID: ***" \ 
  "http://localhost:8000/api/0/relays/projectconfigs/?version=3" | jq '.configs["a4b7b57142ce4fdb8308a7484c2a4250"].config.dynamicSampling'
```

response:
```
      "sampleRate": 0.001,
      "type": "trace",
      "condition": {
        "op": "and",
        "inner": [
          {
            "op": "glob",
            "name": "trace.release",
            "value": [
              "myapp@1.0.1"
            ]
          }
        ]
      },
      "active": true,
      "id": 14
    }
  ]
}
```

### CODE REVIEW

File: src/sentry/models/release.py [hunk 1]
```diff
@@ -67,6 +69,62 @@ class ReleaseCommitError(Exception):
     pass
 
 
+def has_latest_release(rules):
```
[the remark this answers is not in the document]
Comment (untitaker) [reply src/sentry/models/release.py:72]: please wrap this in try-except + capture_exception. I don't think we have enough type safety and schema validation to trust that this function never crashes

[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply src/sentry/models/release.py]: yeah will do.

**Review by matejminar — APPROVED** [review]
Tested the feature, works as expected 👍

Will let others review the actual code.

File: src/sentry/relay/config/__init__.py [hunk 2]
```diff
@@ -179,9 +180,26 @@ def _get_project_config(project, full_config=True, project_keys=None):
...
+                        # get latest overall (no environments filters)
+                        environment = None
+                        rule["condition"]["inner"][0]["value"] = get_latest_release(
+                            [project], environment
+                        )
+                    # end TODO
+                    active_rules.append(rule)
+
+            cfg["config"]["dynamicSampling"] = {"rules": active_rules}
+            # TODO(andrii): schedule_recomputation
```
Comment (untitaker) [inline src/sentry/relay/config/__init__.py:202]: Is that still up to date?

**Review by untitaker — APPROVED** [review]

File: src/sentry/relay/config/__init__.py (hunk 2 above)
Comment (andriisoldatenko) [reply src/sentry/relay/config/__init__.py:202]: oops need to remove it.

Comment (andriisoldatenko) [reply src/sentry/relay/config/__init__.py]: done.

**Review by ahmedetefy — APPROVED** [review]

File: tests/sentry/models/test_release.py [hunk 3]
```diff
@@ -1299,3 +1302,62 @@ def test_simple(self):
         assert Commit.objects.filter(
             id=commit2.id, organization_id=org.id, repository_id=repo.id
         ).exists()
+
+
+from sentry.testutils import TransactionTestCase
+
+
+class ReleaseProjectManagerTestCase(TransactionTestCase):
+    """ """
```
Comment (ahmedetefy) [inline tests/sentry/models/test_release.py:1311]: Was this on purpose?

Comment (andriisoldatenko) [reply tests/sentry/models/test_release.py:1311]: will remove one sec.

File: src/sentry/models/release.py [hunk 4]
```diff
@@ -67,6 +69,75 @@ class ReleaseCommitError(Exception):
     pass
 
 
+def dynamic_sampling_has_latest_release(rules):
```
Comment (ahmedetefy) [inline src/sentry/models/release.py:72]: Nit pick (feel free to ignore)
would rename to `ds_rules_contain_latest_release_rule`

Comment (andriisoldatenko) [reply src/sentry/models/release.py:72]: yeah will do.

**Pull request merged**